cpp/src/arrow/builder-dict.cc

```diff
@@ -33,6 +35,131 @@ namespace arrow {
+      }
+      out_transpose_maps->push_back(std::move(transpose_map));
+    } else {
+      for (int64_t i = 0; i < values.length(); ++i) {
+        memo_table.GetOrInsert(values.GetView(i));
+      }
+    }
+  }
+
+  // Build unified dictionary array
+  std::shared_ptr<ArrayData> data;
+  RETURN_NOT_OK(DictTraits::GetDictionaryArrayData(pool, value_type, memo_table,
+                                                   0 /* start_offset */, &data));
+  *out_values = MakeArray(data);
+  return Status::OK();
+}
+
+Status DictionaryType::Unify(MemoryPool* pool, const std::vector<const DataType*>& types,
+                             std::shared_ptr<DataType>* out_type,
+                             std::vector<std::vector<int32_t>>* out_transpose_maps) {
+  if (types.size() == 0) {
+    return Status::Invalid("need at least one input type");
+  }
+  std::vector<const DictionaryType*> dict_types;
+  dict_types.reserve(types.size());
+  for (const auto& type : types) {
+    if (type->id() != Type::DICTIONARY) {
+      return Status::TypeError("input types must be dictionary types");
+    }
+    dict_types.push_back(checked_cast<const DictionaryType*>(type));
+  }
+
+  // XXX Should we check the ordered flag?
+  auto value_type = dict_types[0]->dictionary()->type();
+  for (const auto& type : dict_types) {
+    auto values = type->dictionary();
+    if (!values->type()->Equals(value_type)) {
+      return Status::TypeError("input types have different value types");
+    }
+    if (values->null_count() != 0) {
+      return Status::TypeError("input types have null values");
+    }
+  }
+
+#define UNIFY_CASE(TYPE_CLASS)                                                     \
+  case TYPE_CLASS::type_id:                                                        \
+    RETURN_NOT_OK(UnifyDictionaryValues<TYPE_CLASS>(pool, value_type, dict_types,  \
+                                                    &values, out_transpose_maps)); \
+    break;
+
+  std::shared_ptr<Array> values;
+  switch (value_type->id()) {
+    UNIFY_CASE(Int8Type)
+    UNIFY_CASE(Int16Type)
+    UNIFY_CASE(Int32Type)
+    UNIFY_CASE(Int64Type)
+    UNIFY_CASE(UInt8Type)
+    UNIFY_CASE(UInt16Type)
+    UNIFY_CASE(UInt32Type)
+    UNIFY_CASE(UInt64Type)
+    UNIFY_CASE(HalfFloatType)
+    UNIFY_CASE(FloatType)
+    UNIFY_CASE(DoubleType)
+    UNIFY_CASE(TimestampType)
+    UNIFY_CASE(Time32Type)
+    UNIFY_CASE(Time64Type)
+    UNIFY_CASE(Date32Type)
+    UNIFY_CASE(Date64Type)
+    UNIFY_CASE(BooleanType)
+    UNIFY_CASE(BinaryType)
+    UNIFY_CASE(StringType)
+    UNIFY_CASE(FixedSizeBinaryType)
+    default: {
+      std::stringstream ss;
+      ss << "Unification of " << value_type->ToString()
+         << " dictionaries is not implemented";
+      return Status::NotImplemented(ss.str());
+    }
+  }
+
+#undef UNIFY_CASE
+
+  // Build unified dictionary type with the right index type
+  std::shared_ptr<DataType> index_type;
+  if (values->length() < 0x7f) {
```

Comment:
use expression with either `INT8_MAX` or `std::numeric_limits...`

Reply:
I don't know, I find it as much or more readable like this.

Reply:
I think this falls under "magic number", though: even if everybody knows what it is and finds it more readable, it's probably better to be explicit with a symbolic constant

Reply:
Ok...